sandbox/libs/authn/docs/openapi.yaml

```diff
@@ -0,0 +1,79 @@
+openapi: 3.0.0
+info:
+  version: 1.0.0
+  title: OpenSearch Identity Permission API
+  license:
+    name: Apache 2.0
+jsonSchemaDialect:
+paths:
+  cluster/permissions:
```

Comment:
In another PR, @DarshitChanpura has introduced a create user API with the endpoint:

```
PUT _identity/api/users/{username} -d {body}
```

Since `username` is a required parameter for granting permissions, what do you think about the route:

```
PUT _identity/api/users/{username}/grant -d {body}
```

Reply:
I can make that change but I have two thoughts in favor of the current state: I don't know how the new version would work for plugins/extensions since you would not have them as part of users. We could make an entirely new action to grant for plugins and extensions but this adds complexity where I do not believe there needs to be. (It also requires adding next route level which we can avoid in the current setup).

I also don't know that the new version would technically be RESTful. The API would not be modifying the user as the resource but instead the permission storage. 

Let me know what you think, I can definitely make the change if that is what you prefer but there was some thought behind the current design so I wanted to explain that thought a bit.

Reply:
In order to re-use the authorization framework being introduced, one idea is that extensions are assigned an identity. One way to accomplish that is to re-use the concept of a user, in this case for an extension, and all of the same endpoints would be applicable to that identifier.

Reply:
Hi Craig, thank you for your reply. 

Treating an extension as a user would let you keep the the `username` route parameter format. That makes sense to me and I will go ahead and make a change to move the principal into the route path. 

I do have concerns about placing everything together under the `users` path or calling the field `username`. In my opinion it is one thing to to put the principal in the route and treat an extension and user similarly behind the scenes but it seems like calling a field `username` and then having it mean `principal` is confusing. Likewise, when we are storing the permissions independent of subject-object I am hesitant to place the route under the `users` path. I know that it may seem like it will not matter to the average user but I just don't want things to be more confusing then they need to be. In my opinion having a "flatter" api library is not worse than a "deeper" one. 

For now, I will go ahead and move the principal into the route to align with what you and @DarshitChanpura are in favor of but I would prefer to not use the term `username` or accessing the api under the `users` route. Maybe we could get the input of @shanilpa for this seeing as he is the UX/UI wizard. Again, I greatly appreciate all your explanation and advice.

Reply:
If `user` sounds too human, then maybe its helpful to think of `accounts`? Its very common for services to run under `service accounts` and regular users to run under `accounts`. They are mostly treated as the same, but one implies the user is a human and the other a service. I still think `username` would be the most apt term for the principal of an account, but open to suggestions.

Reply:
The api won't be under `users` route. Users route is `_identity/api/users/{username}/create`, and Permissions route would be `/_identity/api/permissions/{username}/grant` so a separate tree of routes that start with `/permissions` vs `/users` for User

Reply:
It seems like both of you really prefer `username` so I will go with that. I will split it from the `users` route like DC mentioned however to add a clearer separation.